File: app/src/main/java/igrek/todotree/service/summary/DailySummaryService.java
package igrek.todotree.service.summary;

import android.app.Activity;

import com.google.common.base.Joiner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import igrek.todotree.domain.stats.StatisticEvent;
import igrek.todotree.domain.stats.StatisticEventType;
import igrek.todotree.logger.Logger;
import android.content.Context;
import igrek.todotree.logger.LoggerFactory;
import igrek.todotree.service.statistics.StatisticsLogService;

public class DailySummaryService {
	
	public static final String DAILY_SUMMARY_ACTION = "dailySummary";
	
	private Logger logger = LoggerFactory.getLogger();
	private Context context;
	private NotificationService notificationService;
	private StatisticsLogService statisticsLogService;
	
	public DailySummaryService(Context context, NotificationService notificationService, StatisticsLogService statisticsLogService) {
		this.context = context;
		this.notificationService = notificationService;
		this.statisticsLogService = statisticsLogService;
	}
	
	public void showSummaryNotification() {
		String message = getMessage();
		if (message != null) {
			notificationService.sendNotification(context, "Daily summary", message);
		} else {
			logger.debug("no summary to show");
		}
	}
	
	private String getMessage() {
		try {
			List<StatisticEvent> events = statisticsLogService.getLast24hEvents();
			// build statistics info
			StringBuilder message = new StringBuilder();
			int completed = 0;
			int created = 0;
			// latest first
			Collections.sort(events, (o1, o2) -> o2.getDatetime().compareTo(o1.getDatetime()));
			for (StatisticEvent event : events) {
				if (event.getType().equals(StatisticEventType.TASK_COMPLETED))
					completed++;
				else if (event.getType().equals(StatisticEventType.TASK_CREATED))
					created++;
			}
			
			if (!isSummaryToBeShown(completed, created)) {
				// no message
				return null;
			}
			
			int diff = created - completed;
			
			message.append("You have done a lot today :) (diff: " + diff + ").\n");
			
			message.append("Recently completed tasks (" + completed + "):\n");
			List<String> completedNames = new ArrayList<>();
			for (StatisticEvent event : events) {
				if (event.getType().equals(StatisticEventType.TASK_COMPLETED)) {
					completedNames.add(event.getTaskName());
					//					message.append(" - ").append(datetimeFormat.format(event.getDatetime()));
				}
			}
			Collections.reverse(completedNames);
			message.append(Joiner.on("; ").join(completedNames));
			
			return message.toString();
		} catch (Exception e) {
			logger.error(e);
			return null;
		}
	}
	
	private boolean isSummaryToBeShown(int completed, int created) {
		return completed > created;
	}
}

File: app/src/main/java/igrek/todotree/service/summary/DailySummaryService.kt
package igrek.todotree.service.summary

import android.content.Context
import com.google.common.base.Joiner
import igrek.todotree.info.logger.LoggerFactory.logger
import igrek.todotree.info.logger.Logger.debug
import igrek.todotree.service.statistics.StatisticsLogService.getLast24hEvents
import igrek.todotree.domain.stats.StatisticEvent.datetime
import igrek.todotree.domain.stats.StatisticEvent.type
import igrek.todotree.domain.stats.StatisticEvent.taskName
import igrek.todotree.info.logger.Logger.error
import igrek.todotree.service.summary.NotificationService
import igrek.todotree.service.statistics.StatisticsLogService
import igrek.todotree.info.logger.LoggerFactory
import igrek.todotree.domain.stats.StatisticEvent
import java.lang.StringBuilder
import java.util.Collections
import java.util.Comparator
import igrek.todotree.domain.stats.StatisticEventType
import java.util.ArrayList
import java.lang.Exception

class DailySummaryService(
    private val context: Context,
    private val notificationService: NotificationService,
    private val statisticsLogService: StatisticsLogService
) {
    private val logger = LoggerFactory.logger
    fun showSummaryNotification() {
        val message = message
        if (message != null) {
            notificationService.sendNotification(context, "Daily summary", message)
        } else {
            logger.debug("no summary to show")
        }
    }//					message.append(" - ").append(datetimeFormat.format(event.getDatetime()));// no message

    // build statistics info
    private val message:
    // latest first
            String?
        private get() = try {
            val events = statisticsLogService.getLast24hEvents()
            // build statistics info
            val message = StringBuilder()
            var completed = 0
            var created = 0
            // latest first
            Collections.sort(events) { (_, datetime): StatisticEvent, (_, datetime1): StatisticEvent ->
                datetime1.compareTo(
                    datetime
                )
            }
            for ((type) in events) {
                if (type == StatisticEventType.TASK_COMPLETED) completed++ else if (type == StatisticEventType.TASK_CREATED) created++
            }
            if (!isSummaryToBeShown(completed, created)) {
                // no message
                return null
            }
            val diff = created - completed
            message.append("Recently completed tasks ($completed, diff: $diff):\n")
            val completedNames: MutableList<String?> = ArrayList()
            for ((type, _, taskName) in events) {
                if (type == StatisticEventType.TASK_COMPLETED) {
                    completedNames.add(taskName)
                    //					message.append(" - ").append(datetimeFormat.format(event.getDatetime()));
                }
            }
            Collections.reverse(completedNames)
            message.append(Joiner.on("; ").join(completedNames))
            message.toString()
        } catch (e: Exception) {
            logger.error(e)
            null
        }

    private fun isSummaryToBeShown(completed: Int, created: Int): Boolean {
        return completed > created
    }

    companion object {
        const val DAILY_SUMMARY_ACTION = "dailySummary"
    }
}